### ISSUE TITLE
17.0.6 misc-include-cleaner requires header providing "true"/"false"

### ISSUE THREAD

**Issue by artemisveizi** [llvm/llvm-project#85323]
I think this is a false positive. My understanding is that `true` and `false` are built in to the language, and should not need an explicit #include. I get the following warnings from clang-tidy:
```
pmanager.h:47:24: warning: no header providing "true" is directly included [misc-include-cleaner]
   15 |                 return true;
      |                        ^
```
```
pmanager:50:16: warning: no header providing "false" is directly included [misc-include-cleaner]
   15 |         return false;
      |                ^
```
Just to try and resolve this, I #include <cstdbool>. This does not resolve the warnings above, and instead adds the warning: "`included header cstdbool is not used directly [misc-include-cleaner]`"

This seems like an error, based on my understanding of the documentation.

**Comment by sam-mccall** [comment]
It'd be helpful to have more context+source+logs here. Is this running through clang-include-cleaner, clang-tidy, or clangd?

It's relevant whether pmanager.h is C or C++, and what version, and where the `true` macro is defined in your program.

- if nothing defines `true` as a macro then include-cleaner can't possibly generate that diagnostic
- if something defines `true`...
  - if the file is C++, then that code is invalid - `true` is a keyword
  - if it's C and something other than `stdbool.h` is defining `true` then I think the diagnostic is correct: you're using a macro named `true` and `stdbool.h` is the canonical header you should include for that
  - it's C and `stdbool.h` is defining `true`, then...
    - if it's <C23, then the diagnostic is correct: `true` and `false` are macros, not keywords, and you need to include `<stdbool.h>` (not `<cstdbool>`)
    - if it's C23, then something's wrong with stdbool.h (it's provided by clang, and skips these defines if `__STDC_VERSION__ > 201710L`)
 
Happy to get into more detail once we know which case we're dealing with...
